src/lib/docker-utils.ts

```diff
@@ -225,24 +238,53 @@ export async function fetchDeltaWithProgress(
 							`Got an error when parsing delta server response for v3 delta: ${e}`,
 						);
 					}
-					id = await applyBalenaDelta(name, token, onProgress, logFn);
+					// Try to apply delta DELTA_APPLY_RETRY_COUNT times, then throw DeltaApplyError
+					let tryCount = 0;
+					let lastError;
+					while (tryCount < DELTA_APPLY_RETRY_COUNT) {
+						try {
+							id = await applyBalenaDelta(name, token, onProgress, logFn);
+							break;
+						} catch (e) {
+							tryCount++;
+							lastError = e;
+							logFn(
+								`Delta apply failed, retrying (${tryCount}/${DELTA_APPLY_RETRY_COUNT})...`,
+							);
+						}
+					}
+					if (lastError) {
+						const { statusCode, statusMessage } = lastError as StatusError;
+						throw new DeltaApplyError(statusCode ?? 404, statusMessage);
+					}
```

Comment:
Is better to be strict with typings. Also `for` is preferable here
```suggestion
                                        // Try to apply delta DELTA_APPLY_RETRY_COUNT times, then throw DeltaApplyError
					let lastError: StatusError | undefined = undefined;
					for (
						let tryCount = 0;
						tryCount < DELTA_APPLY_RETRY_COUNT;
						tryCount++
					) {
						try {
							id = await applyBalenaDelta(name, token, onProgress, logFn);
							break;
						} catch (e) {
							if (!isStatusError(e)) {
								throw e;
							}
							lastError = e;
							logFn(
								`Delta apply failed, retrying (${tryCount + 1}/${DELTA_APPLY_RETRY_COUNT})...`,
							);
						}
					}
					if (lastError) {
						const { statusCode, statusMessage } = lastError;
						throw new DeltaApplyError(statusCode, statusMessage);
					}
```